glib/src/value.rs

```diff
@@ -80,44 +76,216 @@
+
+/// A type that can be stored in `Value`s.
+pub trait ValueType: ToValue + for<'r> FromValue<'r> /* TODO: TryFrom<Value> + Into<Value> */ {
+    /// Type to get the `Type` from.
+    ///
+    /// This exists only for handling nullable types.
+    // FIXME: Should default to Self once associated type defaults are stabilized
+    // https://github.com/rust-lang/rust/issues/29661
+    type Type: StaticType;
+}
+
+/// A type that can be stored in `Value`s and is nullable.
+///
+/// These are types were storing an `Option` is valid. Examples are `String` and all object types.
+pub trait ValueTypeNullable:
+    ValueType + ToValueNullable + for<'r> FromValue<'r, Error = WrongValueTypeOrNoneError>
+{
+}
+
+impl<'r, T: ValueTypeNullable> ValueType for Option<T>
+where
+    Self: for<'s> FromValue<'s, Error = WrongValueTypeError>,
+{
+    type Type = T::Type;
+}
+
+/// A type for which a `ParamSpec` exists and that can be used for object properties.
+pub trait HasParamSpec: ValueType {
+    /// Create a new default `ParamSpec` for this type.
+    ///
+    /// For types that allow specifiying a default value or valid range, these will be initialized
+    /// with the equivalent of `None` or `0` and the whole valid range.
+    fn param_spec(name: &str, nick: &str, blurb: &str, flags: ParamFlags) -> ParamSpec;
+}
+
+/// A type for which a `ParamSpec` exists that allows specifying a default value.
+pub trait HasParamSpecWithDefault: HasParamSpec {
+    /// Create a new `ParamSpec` for this type with a default value.
+    ///
+    /// For types that allow specifying a valid range, these will be initialized with the whole
+    /// valid range.
+    fn param_spec_with_default(
+        name: &str,
+        nick: &str,
+        blurb: &str,
+        // FIXME: Should probably be `Self`
+        default: &Self,
+        flags: ParamFlags,
+    ) -> ParamSpec;
+}
+
+/// A type for which a `ParamSpec` exists that allows specifying a default value and valid range.
+pub trait HasParamSpecWithMinMax: HasParamSpecWithDefault {
+    /// Create a new `ParamSpec` for this type.
+    fn param_spec_with_min_max(
+        name: &str,
+        nick: &str,
+        blurb: &str,
+        min: &Self,
+        max: &Self,
+        default: &Self,
+        flags: ParamFlags,
+    ) -> ParamSpec;
+}
 
 /// An error returned from the [`get`](struct.Value.html#method.get)
-/// or [`get_some`](struct.Value.html#method.get_some) functions on a [`Value`](struct.Value.html)
+/// function on a [`Value`](struct.Value.html) for non-nullable types
+/// and `Option`.
 #[derive(Clone, Copy, Debug, Eq, PartialEq)]
-pub struct GetError {
+pub struct WrongValueTypeError {
     pub actual: Type,
     pub requested: Type,
 }
 
-impl GetError {
-    pub fn new_type_mismatch(actual: Type, requested: Type) -> Self {
-        GetError { actual, requested }
+impl WrongValueTypeError {
+    pub fn check<T: StaticType>(value: &Value) -> Result<(), WrongValueTypeError> {
```

Comment:
Something feels wrong with this one. Shouldn't `WrongValueTypeError` result from a `check` somewhere else? Maybe `Value::check_type`?

Reply:
That would make it public `Value` API, which I thought might be confusing for users. But can be done :)

Reply:
Maybe it doesn't need to be made `pub`, if it's only checked from the module / crate.

Reply:
It has to be called by all `FromValue` impls to check the type (in addition to whatever else they might need to check)

Reply:
Indeed. Another solution, if you want to avoid the confusion, would be to add an empty `ValueType`, so we would use `ValueType::check::<T>(value)` but we get turbofished (didn't confirmed), or simply have it on `StaticType` somehow: `StaticType::check(value)`

Reply:
Well, this is basically `Value::is()` (which is implemented in terms of that currently) but for that it wouldn't make sense to return a `Result` :) I'll keep it as it is for now, all other options seem equally strange.